src/main/java/solutions/siren/join/action/coordinate/FilterJoinVisitor.java

```diff
@@ -209,32 +195,83 @@ private void checkForFailure(FilterJoinNode node) {
-    Map<String, Object> binaryFilterParams = new HashMap<>();
-    binaryFilterParams.put("value", bytes);
+    // Create the nested object for the parameters of the field data terms query
+    Map<String, Object> queryParams = new HashMap<>();
+    queryParams.put("value", bytes.bytes);
     // use the hash of the filter join source map as cache key - see #170
-    binaryFilterParams.put("_cache_key", node.getCacheId());
+    queryParams.put("_cache_key", node.getCacheId());
+
+    // Create the nested object for the field
+    Map<String, Object> field = new HashMap<>();
+    field.put(node.getField(), queryParams);
 
-    // Create the object for the filter
-    Map<String, Object> binaryFilter = new HashMap<>();
-    binaryFilter.put(node.getField(), binaryFilterParams);
+    // Create the nested object for the field data terms query
+    Map<String, Object> fieldDataTermsQuery = new HashMap<>();
+    fieldDataTermsQuery.put(FieldDataTermsQueryParser.NAME, field);
 
-    // Add the filter to the parent
-    parent.put(FieldDataTermsQueryParser.NAME, binaryFilter);
+    // Create the object for the constant score query
+    Map<String, Object> constantScoreQueryParams = new HashMap<>();
+    constantScoreQueryParams.put("filter", fieldDataTermsQuery);
+
+    // Add the constant score query to the parent
+    parent.put(ConstantScoreQueryParser.NAME, constantScoreQueryParams);
     node.setState(FilterJoinNode.State.CONVERTED);
     this.blockingQueue.poll();
   }
 
+  protected static class AsyncFilterJoinVisitorAction {
+
+    protected final Client client;
+    protected final FilterJoinNode node;
+    protected final TermsByQueryActionListener listener;
+
+    protected static final ESLogger logger = Loggers.getLogger(AsyncFilterJoinVisitorAction.class);
+
+    protected AsyncFilterJoinVisitorAction(Client client, FilterJoinNode node, TermsByQueryActionListener listener) {
+      this.client = client;
+      this.node = node;
+      this.listener = listener;
+    }
+
+    protected void start() {
+      this.executeTermsByQuery();
+    }
+
+    protected void executeTermsByQuery() {
+      logger.debug("Executing async terms by query action");
+      final TermsByQueryRequest termsByQueryReq = this.getTermsByQueryRequest(node);
+      client.execute(TermsByQueryAction.INSTANCE, termsByQueryReq, listener);
+    }
+
+    protected TermsByQueryRequest getTermsByQueryRequest(FilterJoinNode node) {
+      String[] lookupIndices = node.getLookupIndices();
+      String[] lookupTypes = node.getLookupTypes();
+      String lookupPath = node.getLookupPath();
+      XContentBuilder lookupQuery = node.getLookupQuery();
+      TermsByQueryRequest.Ordering ordering = node.getOrderBy();
+      Integer maxTermsPerShard = node.getMaxTermsPerShard();
+      TermsByQueryRequest.TermsEncoding termsEncoding = node.getTermsEncoding();
+
+      return new TermsByQueryRequest(lookupIndices).field(lookupPath)
```

Comment:
nitpick: put `.field(lookupPath)` on the next line so it is aligned with the rest.